clients/src/main/java/org/apache/kafka/common/requests/DescribeAclsRequest.java

```diff
@@ -18,117 +18,115 @@
-
-        validate(filter, version);
+        this.data = data;
+        validate(version);
     }
 
     public DescribeAclsRequest(Struct struct, short version) {
         super(ApiKeys.DESCRIBE_ACLS, version);
-        ResourcePatternFilter resourceFilter = RequestUtils.resourcePatternFilterFromStructFields(struct);
-        AccessControlEntryFilter entryFilter = RequestUtils.aceFilterFromStructFields(struct);
-        this.filter = new AclBindingFilter(resourceFilter, entryFilter);
+        this.data = new DescribeAclsRequestData(struct, version);
+    }
+
+    public DescribeAclsRequestData data() {
+        return data;
     }
 
     @Override
     protected Struct toStruct() {
-        Struct struct = new Struct(ApiKeys.DESCRIBE_ACLS.requestSchema(version()));
-        RequestUtils.resourcePatternFilterSetStructFields(filter.patternFilter(), struct);
-        RequestUtils.aceFilterSetStructFields(filter.entryFilter(), struct);
-        return struct;
+        return data.toStruct(version());
     }
 
     @Override
     public AbstractResponse getErrorResponse(int throttleTimeMs, Throwable throwable) {
-        return new DescribeAclsResponse(throttleTimeMs, ApiError.fromThrowable(throwable), Collections.emptySet());
+        ApiError error = ApiError.fromThrowable(throwable);
+        DescribeAclsResponseData response = new DescribeAclsResponseData()
+            .setThrottleTimeMs(throttleTimeMs)
+            .setErrorCode(error.error().code())
+            .setErrorMessage(error.message());
+        return new DescribeAclsResponse(response);
     }
 
     public static DescribeAclsRequest parse(ByteBuffer buffer, short version) {
         return new DescribeAclsRequest(ApiKeys.DESCRIBE_ACLS.parseRequest(version, buffer), version);
     }
 
     public AclBindingFilter filter() {
-        return filter;
+        ResourcePatternFilter rpf = new ResourcePatternFilter(
+                ResourceType.fromCode(data.resourceType()),
+                data.resourceNameFilter(),
+                PatternType.fromCode(data.resourcePatternType()));
+        AccessControlEntryFilter acef =  new AccessControlEntryFilter(
+                data.principalFilter(),
+                data.hostFilter(),
+                AclOperation.fromCode(data.operation()),
+                AclPermissionType.fromCode(data.permissionType()));
+        return new AclBindingFilter(rpf, acef);
     }
 
-    private void validate(AclBindingFilter filter, short version) {
-        if (version == 0
-            && filter.patternFilter().patternType() != PatternType.LITERAL
-            && filter.patternFilter().patternType() != PatternType.ANY) {
-            throw new UnsupportedVersionException("Version 0 only supports literal resource pattern types");
+    private void validate(short version) {
+        if (version == 0) {
+            if (data.resourcePatternType() == PatternType.ANY.code()) {
+                data.setResourcePatternType(PatternType.LITERAL.code());
```

Comment:
I tried changing the JSON request definition but I couldn't get it to generate the right Data class so I've had to put this logic :(
As far as I can tell DeleteAclsRequest has the same requirement.